hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMDBTransactionBuffer.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.hadoop.hdds.scm.ha;
+
+import org.apache.hadoop.hdds.scm.metadata.SCMMetadataStore;
+import org.apache.hadoop.hdds.utils.db.BatchOperation;
+import org.apache.hadoop.hdds.utils.db.Table;
+
+import java.io.IOException;
+
+import static org.apache.hadoop.ozone.OzoneConsts.TRANSACTION_INFO_KEY;
+
+public class SCMDBTransactionBuffer {
+  private final SCMMetadataStore metadataStore;
+  private BatchOperation currentBatchOperation;
+  private SCMTransactionInfo latestTrxInfo;
+
+  public SCMDBTransactionBuffer(SCMMetadataStore store) {
+    this.metadataStore = store;
+
+    // initialize a batch operation during construction time
+    currentBatchOperation = this.metadataStore.getStore().initBatchOperation();
+    latestTrxInfo =
+        SCMTransactionInfo
+            .builder()
+            .setTransactionIndex(-1)
+            .setCurrentTerm(-1)
+            .build();
+  }
+
+  public BatchOperation getCurrentBatchOperation() {
+    return currentBatchOperation;
+  }
+
+  public void updateLatestTrxInfo(SCMTransactionInfo info) {
+    this.latestTrxInfo = info;
```

Comment:
Shall we add some checks here to guarantee new info is greater than current lastTrxInfo?

Reply:
Good point. I will add a check or a LOG.error for that case.

Theoretically, the Ratis log will be applied with index monotonically increasing thus the new info is supposed to greater than current  lastTrxInfo. But have a check or a LOG will better handle rare cases or bugs.

Reply:
Done

Reply:
In my opinion, we should stop the ratis server here of such a case arise. It is fatal case where we cannot move ahead from here.

Reply:
Such case should be considered as a bug and be fixed. 

I created a JIRA to further discuss how to deal with such case: https://issues.apache.org/jira/browse/HDDS-4723